ethereum/ratelimit/rate_limiter.go

```diff
@@ -0,0 +1,199 @@
+	// (timePassed / 24hrs) * maxRequestsPer24hrs
+	theoreticalGrantsAccrued := int((float64(timePassedSinceCheckpoint.Nanoseconds()) / float64((24 * time.Hour).Nanoseconds())) * float64(maxRequestsPer24Hrs))
+	// theoreticalGrants - storedGrantedInLast24HrsUTC = accruedGrants
+	accruedGrants := theoreticalGrantsAccrued - storedGrantedInLast24HrsUTC
+
+	// Instantiate limiter with `maxRequestsPer24Hrs` bucketsize and a limit
+	// that results in `maxRequestsPer24Hrs` requests being whitelisted in a 24hr period
+	limit := rate.Limit(float64(maxRequestsPer24Hrs) / (24 * 60 * 60))
+	twentyFourHourLimiter := rate.NewLimiter(limit, maxRequestsPer24Hrs)
+
+	// Since Limiter begins initially full, we drain it before use. i.e., We do not want 100k
+	// requests to already be queued up, instead we only want the number of accrued grants that
+	// have gone unused to be available at startup
+	amountToDrain := maxRequestsPer24Hrs - accruedGrants
+	ctx := context.Background()
+	err = twentyFourHourLimiter.WaitN(ctx, amountToDrain)
+	if err != nil {
+		return nil, err
+	}
+
+	// Instantiate limiter with a bucketsize of one and a limit that results
+	// in no more than `maxRequestsPerSecond` requests per second.
+	limit = rate.Limit(maxRequestsPerSecond)
+	perSecondLimiter := rate.NewLimiter(limit, 1)
+
+	return &RateLimiter{
+		aClock:                aClock,
+		maxRequestsPer24Hrs:   maxRequestsPer24Hrs,
+		twentyFourHourLimiter: twentyFourHourLimiter,
+		perSecondLimiter:      perSecondLimiter,
+		meshDB:                meshDB,
+		currentUTCCheckpoint:  storedUTCCheckpoint,
+		grantedInLast24hrsUTC: storedGrantedInLast24HrsUTC,
+	}, nil
+}
+
+// Start starts two background processes required for the RateLimiter to function. One that
+// stores it's state to the DB at a checkpoint interval, and another that clears accrued
+// grants when the UTC day time window elapses.
+func (r *RateLimiter) Start(ctx context.Context, checkpointInterval time.Duration) error {
+	r.startMutex.Lock()
+	if r.wasStartedOnce {
+		r.startMutex.Unlock()
+		return errors.New("Can only start RateLimiter once per instance")
+	}
+	r.wasStartedOnce = true
+	r.startMutex.Unlock()
+
+	// Start 24hr UTC accrued grants resetter
+	go func() {
+		for {
+			now := r.aClock.Now()
+			currentUTCCheckpoint := getUTCMidnightOfDate(now)
+			nextUTCCheckpoint := time.Date(currentUTCCheckpoint.Year(), currentUTCCheckpoint.Month(), currentUTCCheckpoint.Day()+1, 0, 0, 0, 0, time.UTC)
+			untilNextUTCCheckpoint := nextUTCCheckpoint.Sub(r.aClock.Now())
+			select {
+			case <-ctx.Done():
+				return
+			case <-r.aClock.After(untilNextUTCCheckpoint):
+				// Compute how many grants have accrued and gone unused and remove that
+				// many from the bucket so that it starts empty for the next 24hr period
+				r.mu.Lock()
+				accruedGrants := r.maxRequestsPer24Hrs - r.grantedInLast24hrsUTC
+				if err := r.twentyFourHourLimiter.WaitN(ctx, accruedGrants); err != nil {
```

Comment:
The documentation does not specify whether calling`AllowN` will deduct _however many_ tokens have accrued before returning `false`. I dug into the code and if there aren't `N` tokens that have accrued, the request is rejected and the bucket is not emptied. We must use `WaitN`.

Reply:
Hmm.. that's really annoying. @fabioberger are we confident that there is no situation where this could block?

Reply:
Yes, I am confident this should never block. When we instantiate the rate limiter, we drain the bucket so that the number of accrued grants is `theoreticalGrantsAccrued` - `grantedInLast24hrsUTC` (where the theoretical grants accrued is calculated from the time elapsed since the start of the UTC day interval). Therefore the grants accrued left to be drained will be `maxRequestsPer24Hrs` -  `grantedInLast24hrsUTC`.